### PR TITLE
allow swap key

### PR DESCRIPTION
**Pull request by sj6219**
I implemented #2365

### CODE REVIEW

**Comment by rustdesk** [comment]
@fufesou please review.

**Comment by fufesou** [comment]
@sj6219 Hi, I've briefly read the code. The implementation seems good.

1. Would please run cargo format to the files you've modified?
2. Please check remote peer OS. If MacOS --> MacOS, do not show the swap option.

**Comment by fufesou** [comment]
It's not convenient for me to test right now, I will test it tomorrow.

**Comment by sj6219** [comment]
> 2\. For connections. I think it is necessary to check remote peer OS . If MacOS --> MacOS, do not swap key.

If both computers use swap settings, there is no effect because key input is swapped  once on the client,  and the server swapped it again.

**Comment by fufesou** [comment]
> > 2. For connections. I think it is necessary to check remote peer OS . If MacOS --> MacOS, do not swap key.
> 
> If both computers use swap settings, there is no effect because key input is swapped once on the client, and the server swapped it again.

Yes. But it will go strange if only one side use swap setting, which is possible.
Check if both MacOS and ignore swapping is what we can control, and the result will always be expected.

**Comment by fufesou** [comment]
I think it may be better to allow swap key on the local side, whether it's MacOS or not.
The remote side may not swap keys.

It will be more convenient if the control side can decide whether or not to swap keys.

**Comment by fufesou** [comment]
> To do so, We need to save settings for each server the client access. It will take time to change it because I am a beginner in flutter language.

Maybe no need to save settings for each connection.

The setting can be what you've designed.
1. It should checked and applied when each connection is established.
2. It should be applied to each establised connection when the setting is changed.

BTW, "Allow swap key" option may be easier to find on the remote page(remote menu). Like the [`kb_layout_type`](https://github.com/fufesou/rustdesk/blob/85dc3b0e586fa8ca35ac729fb03fd71019f9772a/libs/hbb_common/src/config.rs#L1074) option.

**Comment by sj6219** [comment]
I tested it on windows and macos.

I didn't implement Linux because I didn't have a development environment.

And, there seems to be a bug in the old non flutter version.
The check box does not appear as shown below.
sciter

**Comment by fufesou** [comment]
Hi, @sj6219  the option is under the keyboard button.

File: src/keyboard.rs [hunk 1]
```diff
@@ -205,11 +205,63 @@ static mut IS_0X021D_DOWN: bool = false;
 pub fn start_grab_loop() {
     #[cfg(any(target_os = "windows", target_os = "macos"))]
     std::thread::spawn(move || {
-        let try_handle_keyboard = move |event: Event, key: Key, is_press: bool| -> Option<Event> {
+            let try_handle_keyboard = move |mut event: Event, key: Key, is_press: bool| -> Option<Event> {
             // fix #2211：CAPS LOCK don't work
             if key == Key::CapsLock || key == Key::NumLock {
                 return Some(event);
             }
+            {
```
Comment (rustdesk) [inline src/keyboard.rs]: Can we create a function for this part of code?

**Review by rustdesk — CHANGES_REQUESTED** [review]

File: src/keyboard.rs (hunk 1 above)
Comment (sj6219) [reply src/keyboard.rs]: I made a block to distinguish it from the changed part. You can get rid of it.

Comment (rustdesk) [reply src/keyboard.rs]: A new function is better, no need to put everything in one function.

Comment (rustdesk) [reply src/keyboard.rs]: Which is bad coding practise.

**Comment by rustdesk** [comment]
Thanks for your change.
@fufesou  I have another concern, this can not work in Wayland AppInage / Flatpak version which get input from flutter rather than via grab system input.

**Comment by sj6219** [comment]
The current implementation swaps data as soon as client receive the keystroke. Would it be simpler to swap data just before it's sent to the server?

https://github.com/rustdesk/rustdesk/blob/master/src/ui_session_interface.rs#:~:text=msg_out.set_key_event(evt.clone())%3B

**Comment by rustdesk** [comment]
AnyDesk does not have the choice to allow swap key.

https://support.anydesk.com/knowledge/anydesk-on-macos#cross-plattform-shortcuts

So i start to hesitate.

**Comment by rustdesk** [comment]
> Would it be simpler to swap data just before it's sent to the server?

Yes

**Comment by rustdesk** [comment]
> AnyDesk does not have the choice to allow swap key.
> 
> https://support.anydesk.com/knowledge/anydesk-on-macos#cross-plattform-shortcuts
> 
> So i start to hesitate.

Appologize, I find where to modify swap on AnyDesk.

**Comment by rustdesk** [comment]
image
It is a global setting on controlling side, effective for outgoing connections.

**Comment by sj6219** [comment]
I'm fixing bug. Please wait.

**Comment by sj6219** [comment]
Bug fixed

**Comment by achimmihca** [comment]
Thanks for implementing this. I think this is what I am looking for.
I use a Windows host to connect to macOS and want to use for example `ctrl+c` to copy text.
On macOS I installed [Karabiner](https://karabiner-elements.pqrs.org/) for this, which works nicely but not when connected via RustDesk.

So again, thanks for implementing this option.

File: Cargo.toml [hunk 2]
```diff
@@ -167,3 +167,8 @@ panic = 'abort'
 strip = true
 #opt-level = 'z' # only have smaller size after strip
 rpath = true
+
+[patch."https://github.com/fufesou/rdev"]
+#rdev = { path = "../rdev" }
+rdev = { git = "https://github.com/sj6219/rdev", branch = "sigma" }
+
```
Comment (rustdesk) [inline Cargo.toml]: Please submit a PR to fufesou/rdev.

File: flutter/lib/desktop/widgets/remote_menubar.dart [hunk 3]
```diff
@@ -1580,6 +1580,9 @@ class _RemoteMenubarState extends State<RemoteMenubar> {
         ),
       );
     }
+    keyboardMenu.add(_createSwitchMenuEntry(
+        'Swap Control-Command  Key', 'allow_swap_key', EdgeInsets.zero, true));
+
     return keyboardMenu;
```
Comment (rustdesk) [inline flutter/lib/desktop/widgets/remote_menubar.dart]: Please do this only if the peer os is mac.

File: src/ui/header.tis [hunk 4]
```diff
@@ -156,6 +156,7 @@ class Header: Reactor.Component {
             <menu.context #keyboard-options>
                 <li #legacy><span>{svg_checkmark}</span>{translate('Legacy mode')}</li> 
                 <li #map><span>{svg_checkmark}</span>{translate('Map mode')}</li> 
+                <li #allow_swap_key><span>{svg_checkmark}</span>{translate('Swap Control-Command Key')}</li>
```
Comment (rustdesk) [inline src/ui/header.tis]: Only if peer os is mac.

File: src/ui_session_interface.rs [hunk 5]
```diff
@@ -843,6 +920,22 @@ impl<T: InvokeUiSession> Interface for Session<T> {
             handle_test_delay(t, peer).await;
         }
     }
+    fn swap_modifier_mouse(&self, msg : &mut hbb_common::protos::message::MouseEvent) {
```
Comment (rustdesk) [inline src/ui_session_interface.rs:1015]: miss new line above.

**Review by rustdesk — CHANGES_REQUESTED** [review]

File: flutter/lib/desktop/widgets/remote_menubar.dart (hunk 3 above)
Comment (rustdesk) [reply flutter/lib/desktop/widgets/remote_menubar.dart]: and local side is not mac

**Comment by sj6219** [comment]
I was hospitalized because I was sick. 
So I can't do coding work for 1-2 weeks.
 I want someone else to work or wait for me.

**Comment by rustdesk** [comment]
Take care, this is not urgent.

**Review by IRZTAR — COMMENTED** [review]
si.

File: Cargo.toml (hunk 2 above)
Comment (rustdesk) [reply Cargo.toml]: Could you please solve this?

Comment (sj6219) [reply Cargo.toml]: I submitted a PR, but my PR has not yet been accepted.

File: flutter/lib/desktop/widgets/remote_menubar.dart (hunk 3 above)
Comment (rustdesk) [reply flutter/lib/desktop/widgets/remote_menubar.dart]: This seems not solved yet.

**Review by rustdesk — CHANGES_REQUESTED** [review]

File: flutter/lib/desktop/widgets/remote_menubar.dart (hunk 3 above)
Comment (rustdesk) [reply flutter/lib/desktop/widgets/remote_menubar.dart]: Ignore this, solved.

File: Cargo.toml (hunk 2 above)
Comment (fufesou) [reply Cargo.toml]: Sorry for late, the PR is accepted now.

Comment (rustdesk) [reply Cargo.toml]: Thanks, let me notify him.

**Comment by rustdesk** [comment]
Do we need to update Cargo.lock for rdev?

**Comment by rustdesk** [comment]
> Do we need to update Cargo.lock for rdev?

CI failed because of this i guess.

**Comment by rustdesk** [comment]
Let me do this.

**Comment by rustdesk** [comment]
@sj6219 do you have interest to work on this https://github.com/rustdesk/rustdesk/issues/230#issuecomment-1422615080?

**Comment by ylluminate** [comment]
Any news here? Very problematic when trying to interact with Control Remote Desktop and cmd-a + cmd-z / cmd-shift-z don't work:
image

**Comment by discoveryX64** [comment]
My issue is completely different from this, swap key is off, and even with it on beforehand it worked perfectly fine

**Comment by linnv** [comment]
how about swap  Caps  Lock <--> Control?

**Comment by linnv** [comment]
> > To do so, We need to save settings for each server the client access. It will take time to change it because I am a beginner in flutter language.
> 
> Maybe no need to save settings for each connection.
> 
> The setting can be what you've designed. It is an option that should be set once and applied to each connection.
> 
> 1. It should checked and applied when each connection is established.
> 2. It should be applied to each establised connection when the setting is changed.
> 
> BTW, "Allow swap key" option may be easier to find on the remote page(remote menu). Like the [`kb_layout_type`](https://github.com/fufesou/rustdesk/blob/85dc3b0e586fa8ca35ac729fb03fd71019f9772a/libs/hbb_common/src/config.rs#L1074) option.

Swap control--command key works on win->mac(windows control mac), nice feature!
how about swap Caps Lock <--> Control?

**Pull request merged**